File: tests/cs-sdk/TestSignalSubscribeIntegrationLoop.cs
// loop test: Signal-Subscribe (Sender + Subscriber)
using System;
using System.Threading;
using System.Threading.Tasks;
using KnotLink;

const int MAX_ITERATIONS = 5;
var cts = new CancellationTokenSource();

var t = Task.Run(() => RunSubscriber(cts.Token));
await Task.Delay(100);
await RunSender(cts.Token);
cts.Cancel();
await t;

static async Task RunSubscriber(CancellationToken ct)
{
    var sub = new SignalSubscriber("app.knotlinksdk.test", "test_signal")
    {
        OnSignalAsync = data =>
        {
            Console.WriteLine($"[Subscriber] Received: {data}");
            return Task.CompletedTask;
        }
    };
    Console.WriteLine("[Subscriber] Listening...");
    while (!ct.IsCancellationRequested)
        await Task.Delay(1000, ct);
}

static async Task RunSender(CancellationToken ct)
{
    await Task.Delay(1000, ct);
    var sender = new SignalSender("app.knotlinksdk.test", "test_signal");
    try
    {
        for (int i = 1; i <= MAX_ITERATIONS && !ct.IsCancellationRequested; i++)
        {
            var data = $"Signal #{i}";
            await sender.EmitAsync(data);
            Console.WriteLine($"[Sender] Sent: {data}");
            await Task.Delay(3000, ct);
        }
    }
    finally { sender.Dispose(); }
}

File: knotlink-sdk-cpp/tests/test_signal_subscribe_integration_loop.cpp
// loop test: Signal-Subscribe (Sender + Subscriber)
#include "../knotlink/SignalSubscriber.hpp"
#include "../knotlink/SignalSender.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <atomic>
using namespace knotlink;

static std::atomic<bool> shutdown{false};
constexpr int MAX_ITERATIONS = 5;

void run_subscriber() {
    SignalSubscriber sub("app.knotlinksdk.test", "test_signal");
    sub.setOnDataReceivedCallback([](const std::string& data) {
        std::cout << "[Subscriber] Received: " << data << std::endl;
    });
    std::cout << "[Subscriber] Listening..." << std::endl;
    while (!shutdown) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

void run_sender() {
    std::this_thread::sleep_for(std::chrono::seconds(1));
    SignalSender sender;
    sender.setConfig("app.knotlinksdk.test", "test_signal");
    for (int i = 1; i <= MAX_ITERATIONS && !shutdown; i++) {
        std::string data = "Signal #" + std::to_string(i);
        sender.emitt(data);
        std::cout << "[Sender] Sent: " << data << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(3));
    }
    shutdown = true;
}

int main() {
    std::thread t(run_subscriber);
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    run_sender();
    t.join();
    return 0;
}
